src/VisualStudio/Core/Def/Implementation/PullMemberUp/MainDialog/PullMemberUpDialogViewModel.cs

```diff
@@ -0,0 +1,216 @@
+        }
+
+        public void SelectPublicMembers()
+        {
+            SelectMembers(Members.WhereAsArray(memberViewModel => memberViewModel.Symbol.DeclaredAccessibility == Accessibility.Public));
+        }
+
+        public void EnableOrDisableOkButton()
+        {
+            var selectedMembers = Members
+                .WhereAsArray(memberSymbolView => memberSymbolView.IsChecked && memberSymbolView.IsCheckable);
+            OkButtonEnabled = SelectedDestination != null && selectedMembers.Count() != 0;
+        }
+
+        public void CheckAndSetStateOfSelectAllCheckBox()
+        {
+            var checkableMembers = Members.WhereAsArray(member => member.IsCheckable);
+            if (checkableMembers.All(member => member.IsChecked))
+            {
+                SelectAllCheckBoxState = true;
+                SelectAllCheckBoxThreeStateEnable = false;
+            }
+            else if (checkableMembers.Any(member => member.IsChecked))
+            {
+                SelectAllCheckBoxThreeStateEnable = true;
+                SelectAllCheckBoxState = null;
+            }
+            else
+            {
+                SelectAllCheckBoxState = false;
+                SelectAllCheckBoxThreeStateEnable = false;
+            }
+        }
+
+        public void SelectDependents()
+        {
+            var checkedMembers = Members
+                .WhereAsArray(member => member.IsChecked && member.IsCheckable);
+
+            var waitResult = _waitIndicator.Wait(
+                    title: ServicesVSResources.Pull_Members_Up,
+                    message: ServicesVSResources.Calculating_dependents,
+                    allowCancel: true,
+                    showProgress: true,
+                    context =>
+                    {
+                        foreach (var member in Members)
+                        {
+                            _symbolToDependentsMap[member.Symbol].Wait(context.CancellationToken);
+                        }
+                    });
+
+            if (waitResult == WaitIndicatorResult.Completed)
+            {
+                foreach (var member in checkedMembers)
+                {
+                    var membersToSelected = FindDependentsRecursively(member.Symbol).SelectAsArray(symbol => _symbolToMemberViewMap[symbol]);
+                    SelectMembers(membersToSelected);
+                }
+            }
+        }
+
+        public void DeSelectAllMembers()
+        {
+            foreach (var member in Members.WhereAsArray(viewModel => viewModel.IsCheckable))
+            {
+                member.IsChecked = false;
+            }
+
+            CheckAndSetStateOfSelectAllCheckBox();
+            EnableOrDisableOkButton();
+        }
+
+        private ImmutableHashSet<ISymbol> FindDependentsRecursively(ISymbol member)
+        {
+            var queue = new Queue<ISymbol>();
+            var visited = new HashSet<ISymbol>();
+            var result = new HashSet<ISymbol>();
```

Comment:
I use visited to avoid the loop in the graph, if method A() call method B(), then B() also called A(), I think it will be a infinite loop without visited